nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -3721,4 +3811,352 @@ pub mod test {
+            )
+            .expect("failed to add nexus zone");
+
+        let blueprint2 = builder.build();
+        verify_blueprint(&blueprint2);
+
+        // Find all nexus zones and verify they have the same generation
+        let nexus_zones: Vec<_> = blueprint2
+            .all_omicron_zones(BlueprintZoneDisposition::any)
+            .filter_map(|(_, zone)| match &zone.zone_type {
+                BlueprintZoneType::Nexus(nexus) => Some(nexus),
+                _ => None,
+            })
+            .collect();
+
+        assert_eq!(nexus_zones.len(), 2);
+        assert_eq!(nexus_zones[0].nexus_generation, existing_nexus_gen);
+        assert_eq!(nexus_zones[1].nexus_generation, existing_nexus_gen);
+
+        logctx.cleanup_successful();
+    }
+
+    /// Test nexus generation assignment logic for different image sources
+    #[test]
+    fn test_nexus_generation_assignment_different_image_increment() {
+        static TEST_NAME: &str =
+            "test_nexus_generation_assignment_different_image_increment";
+        let logctx = test_setup_log(TEST_NAME);
+        let mut rng = SimRngState::from_seed(TEST_NAME);
+
+        // Start with a system that has one Nexus zone
+        let (example_system, blueprint) =
+            ExampleSystemBuilder::new(&logctx.log, TEST_NAME)
+                .nexus_count(1)
+                .build();
+        verify_blueprint(&blueprint);
+
+        // Get the generation of the existing nexus zone
+        let existing_nexus_gen = blueprint
+            .all_omicron_zones(BlueprintZoneDisposition::any)
+            .find_map(|(_, zone)| match &zone.zone_type {
+                BlueprintZoneType::Nexus(nexus) => {
+                    assert_eq!(
+                        zone.image_source,
+                        BlueprintZoneImageSource::InstallDataset
+                    );
+                    Some(nexus.nexus_generation)
+                }
+                _ => None,
+            })
+            .expect("should have found existing nexus");
+
+        let mut builder = BlueprintBuilder::new_based_on(
+            &logctx.log,
+            &blueprint,
+            &example_system.input,
+            &example_system.collection,
+            "test",
+            rng.next_planner_rng(),
+        )
+        .expect("failed to create builder");
+
+        // Get a different sled
+        let sled_ids: Vec<_> = example_system
+            .input
+            .all_sled_ids(SledFilter::Commissioned)
+            .collect();
+        let second_sled_id = sled_ids[1];
+
+        // Use a different image source (artifact vs install dataset)
+        let different_image_source = BlueprintZoneImageSource::Artifact {
+            version: BlueprintArtifactVersion::Available {
+                version: ArtifactVersion::new_const("1.2.3.4"),
+            },
+            hash: ArtifactHash([0x42; 32]),
+        };
+
+        // Add another Nexus zone with different image source - should increment generation
```

Comment:
Since `test_nexus_generation_assignment_new_generation()` already tested that if you pass a particular generation to `sled_add_zone_nexus()` then you get a zone with that generation, I wonder if we should just have this test call `determine_nexus_generation()` with both cases: one with the same image and one with a new one.  I don't think we really need to add the zones and check that they got what we expected.